fdbserver/RestoreMaster.actor.h

```diff
@@ -90,98 +101,188 @@ struct RestoreMasterData : RestoreRoleData, public ReferenceCounted<RestoreMaste
-				ASSERT(prevVersion <= rangeFile.beginVersion);
-				prevVersion = rangeFile.beginVersion;
-				ASSERT(fIndexSet.find(rangeFile.fileIndex) == fIndexSet.end());
-				fIndexSet.insert(rangeFile.fileIndex);
+
+			double nextVersionSize;
+			int nextRangeIdx;
+			std::vector<RestoreFileFR> curLogFiles;
+			std::tie(nextVersionSize, nextRangeIdx, curLogFiles) =
+			    getVersionSize(prevEndVersion, nextVersion, rangeFiles, rangeIdx, logFiles);
+
+			TraceEvent("FastRestoreBuildVersionBatch")
+			    .detail("VersionBatchBeginVersion", vb.beginVersion)
+			    .detail("PreviousEndVersion", prevEndVersion)
+			    .detail("NextVersion", nextVersion)
+			    .detail("RangeIndex", rangeIdx)
+			    .detail("RangeFiles", rangeFiles.size())
+			    .detail("LogIndex", logIdx)
+			    .detail("LogFiles", logFiles.size())
+			    .detail("BatchSizeThreshold", opConfig.batchSizeThreshold)
+			    .detail("CurrentBatchSize", vb.size)
+			    .detail("NextVersionIntervalSize", nextVersionSize)
+			    .detail("NextRangeIndex", nextRangeIdx)
+			    .detail("UsedLogFiles", curLogFiles.size());
+
+			if (vb.size + nextVersionSize <= opConfig.batchSizeThreshold) {
+				// nextVersion should be included in this batch
+				bool advanced = false;
+				vb.size += nextVersionSize;
+				while (rangeIdx < nextRangeIdx) {
+					vb.rangeFiles.push_back(rangeFiles[rangeIdx]);
+					++rangeIdx;
+					advanced = true;
+				}
+
+				for (auto& log : curLogFiles) {
+					ASSERT(log.beginVersion < nextVersion);
+					ASSERT(log.endVersion > prevEndVersion);
+					vb.logFiles.push_back(log);
+					advanced = true;
+				}
+
+				ASSERT(advanced == true || nextVersionSize > 0); // Ensure progress
+				vb.endVersion = nextVersion;
+				prevEndVersion = vb.endVersion;
+			} else {
+				if (vb.size < 1) {
+					// [vb.endVersion, nextVersion) > opConfig.batchSizeThreshold. We should split the version range
+					if (prevEndVersion >= nextVersion) {
+						// If range files at one version > batchSizeThreshold, DBA should increase batchSizeThreshold
+						TraceEvent(SevError, "FastRestoreBuildVersionBatch")
+						    .detail("NextVersion", nextVersion)
+						    .detail("PreviousEndVersion", prevEndVersion)
+						    .detail("NextVersionIntervalSize", nextVersionSize)
+						    .detail("BatchSizeThreshold", opConfig.batchSizeThreshold);
+					}
+					ASSERT(prevEndVersion < nextVersion); // Ensure progress
+					nextVersion = (prevEndVersion + nextVersion) / 2;
+					rewriteNextVersion = true;
+					TraceEvent("FastRestoreBuildVersionBatch")
+					    .detail("NextVersionIntervalSize", nextVersionSize); // Duplicate Trace
+					continue;
+				}
+				// Finalize the current version batch
+				versionBatches->emplace(vb.beginVersion, vb); // copy vb to versionBatch
+				// start finding the next version batch
+				vb.reset();
+				vb.size = 0;
+				vb.beginVersion = prevEndVersion;
+				nextVersion = prevEndVersion + 1;
```

Comment:
Should this be setting `rewriteNextVersion`? Otherwise, I think `nextVersion` just gets increased again in the next iteration.

Reply:
We are supposed to increase the `nextVersion` in the next iteration until the data in [beginVersion,  `nextVersion`) is larger than the threshold. 
`rewriteNextVersion `  is more like `nextVersionHasBeenCalculated`.  It is used only when the current working version interval [`prevEndVersion`,  `nextVersion`) cause the batchSize larger than the threshold. We want to binary search the `nextVersion` when the batchSize is no larger than the threshold.

This logic is in L255 to L271

Reply:
I'm maybe a little confused, but it looks like this is computing `nextVersion=prevEndVersion+1`, such that you would need to say `nextVersionHasBeenCalculated` is true. Otherwise, I'm not sure the setting of `nextVersion` here does anything because it will get reset the next iteration.

Reply:
Oh, my bad. I was thinking something else.

The reason why we need to calculate `nextVersion = prevEndVersion+1` is because:
say we have range file [1] at version 100 and range file [2] at version 101.  When this new batch starts, the `rangeIdx = 2` and  `vb.beginVersion = 101`. If we use the logic at the beginning `nextVersion = std::max(rangeFiles[rangeIdx].version + 1, nextVersion);` without setting `nextVersion` here, the new batch may have version range [101, 101).

Reply:
It looks like `prevEndVersion` is set only at line 253 to be equal to `nextVersion`, and when it gets set at that line, I think `nextVersion` will in the very next iteration be increased to at least 1 larger than its current value (if the sort of `logFiles` is changed to use `endVersion`). I think this is true even if we do the bisection and get no range files because the bisection would have decreased `nextVersion` such that it would be increased again when we finally go around the loop normally. It's possible I'm missing some case though.

If what I said is true and the range you are referring to is `[prevEndVersion, nextVersion)`, it would seem then that it's not possible for this  range to be empty.

Reply:
About ` I think nextVersion will in the very next iteration be increased to at least 1 larger than its current value (if the sort of logFiles is changed to use endVersion). `,
Do you mean Line 191 `nextVersion = std::max(rangeFiles[rangeIdx].version + 1, nextVersion);` will increase `nextVersion`?

If so, the corner case that makes `nextVersion` in next iteration not increase by 1 if the current `nextVersion ` is decided by `rangeFiles[rangeIdx].version + 1` and `rangeIdx` is not changing in the current iteration.

Reply:
Can `rangeIdx` stay the same if we don't do the bisection? In the case of the bisection, we would have also decreased `nextVersion`, so the next iteration will increase it again.

Reply:
Yes, it is possible `rangeIdx` stay the same. It is also possible there are multiple range files at the same version.